core/isolate.rs

```diff
@@ -12,42 +12,87 @@ use crate::libdeno::Snapshot1;
 use crate::libdeno::Snapshot2;
 use crate::shared_queue::SharedQueue;
 use crate::shared_queue::RECOMMENDED_SIZE;
-use futures::Async;
+use futures::stream::{FuturesUnordered, Stream};
+use futures::task;
+use futures::Async::*;
 use futures::Future;
 use futures::Poll;
 use libc::c_void;
-use std::collections::VecDeque;
 use std::ffi::CStr;
 use std::ffi::CString;
+use std::fmt::{Debug, Formatter, Result as FmtResult};
 use std::ptr::null;
 use std::sync::{Arc, Mutex, Once, ONCE_INIT};
 
 pub type Buf = Box<[u8]>;
-pub type Op = dyn Future<Item = Buf, Error = ()> + Send;
 
+pub struct Tagged<F> {
+  future: F,
+  tag: &'static str,
+}
+impl<F> Future for Tagged<F>
+where
+  F: Future,
+{
+  type Item = F::Item;
+  type Error = F::Error;
+  fn poll(&mut self) -> Poll<F::Item, F::Error> {
+    self.future.poll()
+  }
+}
+impl<F> Debug for Tagged<F> {
+  fn fmt(&self, f: &mut Formatter) -> FmtResult {
+    f.write_str(self.tag)
+  }
+}
+
+pub trait Tag
+where
+  Self: Future + Sized,
+{
+  fn tag(self, tag: &'static str) -> Tagged<Self> {
+    Tagged { future: self, tag }
+  }
+}
+impl<T> Tag for T where T: Future + Sized {}
+
+pub trait Op: Future<Item = Buf, Error = ()> + Debug + Send {
+  fn box_op(self) -> Box<dyn Op>;
+}
+impl<T> Op for T
+where
+  T: Future<Item = Buf, Error = ()> + Debug + Send + Sized + 'static,
+{
+  fn box_op(self) -> Box<dyn Op> {
+    Box::new(self)
+  }
+}
+
+#[derive(Debug)]
 struct PendingOp {
   op: Box<Op>,
-  polled_recently: bool,
   zero_copy_id: usize, // non-zero if associated zero-copy buffer.
 }
 
+struct OpResult {
```

Comment:
Hmm - "OpResult" isn't such a good name because it reminds me of "Result"
"BufAndZeroCopyId" is not great either...